From d1a8efee5fef6cd1a359063a1997777ee2ef1e52 Mon Sep 17 00:00:00 2001
From: Michel Thierry <michel.thierry@intel.com>
Date: Fri, 20 Nov 2015 12:10:50 +0000
Subject: [PATCH 28/28] SQUASHME! [VPG]: drm/i915: Prevent wa_ctx obj leak in
 older kernels

In kernel v3.14, i915_gem_init_hw re-inits all the rings, which will
create a new wa_ctx obj. The original upstreamed patch didn't require
such check because newer kernels only init engines once [1].

SQUASHME! - This patch should be squashed into the following existing patch:

	Author: Arun Siluvery <arun.siluvery@linux.intel.com>
	Date: Jun 19, 2015 7:07 PM
	Change-Id: If4c1f74db011d2393f7d5057e04d8b5b4f8aa223
	FROM_UPSTREAM [VPG]: drm/i915/gen8: Add infrastructure to initialize WA
	batch buffers

[1] http://cgit.freedesktop.org/drm-intel/commit/?id=35a57ffbb10840af219eeaf64718434242bb7c76

Change-Id: I0da924afbf284e59e46420309064d29c67acb2a6
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9535
Signed-off-by: Lei Shen <lei.shen@intel.com>
Signed-off-by: Michel Thierry <michel.thierry@intel.com>
(cherry picked from commit 65c196935fc768c365e1879db8fc13648a6c4c99)
Reviewed-on: https://android.intel.com:443/445257
---
 drivers/gpu/drm/i915/intel_lrc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_lrc.c b/drivers/gpu/drm/i915/intel_lrc.c
index 7d4f687..92f47ed 100644
--- a/drivers/gpu/drm/i915/intel_lrc.c
+++ b/drivers/gpu/drm/i915/intel_lrc.c
@@ -2337,6 +2337,15 @@ static int lrc_setup_wa_ctx_obj(struct intel_engine_cs *ring, u32 size)
 {
 	int ret;
 
+	/*
+	 * In more recent kernels, rings are only initialized once and there
+	 * was no risk of leaking the wa_ctx obj. But this is not the case in
+	 * v3.14, i915_resume will re-init the rings (gt.init_rings), so it
+	 * isn't required to create a new wa_ctx obj.
+	 */
+	if (ring->wa_ctx.obj)
+		return 0;
+
 	ring->wa_ctx.obj = i915_gem_alloc_object(ring->dev, PAGE_ALIGN(size));
 	if (!ring->wa_ctx.obj) {
 		DRM_DEBUG_DRIVER("alloc LRC WA ctx backing obj failed.\n");
-- 
1.9.1

